### ISSUE TITLE
Duplicating a package repo package causes config to silently become invalid

### ISSUE THREAD

**Issue by arvindsv** [gocd/gocd#7974]
##### Issue Type

- Bug Report

##### Summary

Duplicating a package repo package causes config to silently become invalid.

###### Basic environment details

* Go Version: `20.3.0 (11467-1fb7c3cc3a71abb71e827f5bbdc2cb139edcd660)`

##### Steps to Reproduce

1. Create a package inside a package repository. You can use this config, with the [maven-repo plugin](https://github.com/1and1/go-maven-poller/releases/tag/go-maven-poller-1.5.1):

    ```
    <repository id="4a6d9a76-0bcd-4858-8c6b-f3f956b938bd" name="repo1">
      <pluginConfiguration id="maven-repo" version="" />
      <configuration>
        <property>
          <key>REPO_URL</key>
          <value>https://repo1.maven.org/maven2</value>
        </property>
        <property>
          <key>USERNAME</key>
          <value />
        </property>
        <property>
          <key>PASSWORD</key>
          <encryptedValue />
        </property>
        <property>
          <key>PROXY</key>
          <value />
        </property>
        <property>
          <key>TIME_ZONE</key>
          <value />
        </property>
        <property>
          <key>LATEST_VERSION_TAG</key>
          <value />
        </property>
      </configuration>
      <packages>
        <package id="7c04de28-6094-400d-b1d4-84dac0cc9c40" name="package1" autoUpdate="false">
          <configuration>
            <property>
              <key>GROUP_ID</key>
              <value>cd.go.plugin</value>
            </property>
            <property>
              <key>ARTIFACT_ID</key>
              <value>go-plugin-api</value>
            </property>
            <property>
              <key>PACKAGING</key>
              <value>jar</value>
            </property>
            <property>
              <key>POLL_VERSION_FROM</key>
              <value />
            </property>
            <property>
              <key>POLL_VERSION_TO</key>
              <value />
            </property>
          </configuration>
        </package>
      </packages>
    </repository>
    ```

2. Duplicate **the package** using the SPA, and give it a different name. It succeeds.

The config is now invalid. If you restart the server, it will not come back up. You can also verify this by editing the config XML and trying to save it.

##### Any other info

The error shown when you save the config is:

```
Cannot save package or repo, found duplicate packages. [Repo Name: 'repo1', Package Name: 'package1'], [Repo Name: 'repo1', Package Name: 'p1']
```

**Comment by maheshp** [comment]
I took a look at this issue, the reason for this behavior is related to how the XML or JSON is deserialized to ConfigurationProperty object and the corresponding fingerprint generation.  

```xml
  <configuration>
        <property>
          <key>GROUP_ID</key>
          <value>cd.go.plugin</value>
        </property>
        <property>
          <key>ARTIFACT_ID</key>
          <value>go-plugin-api</value>
        </property>
        <property>
          <key>PACKAGING</key>
          <value>jar</value>
        </property>
        <property>
          <key>POLL_VERSION_FROM</key>
          <value />
        </property>
        <property>
          <key>POLL_VERSION_TO</key>
          <value />
        </property>
      </configuration>
```

The string computed to generate fingerprint:
```
plugin-id=maven-repo<|>GROUP_ID=cd.go.plugin<|>ARTIFACT_ID=go-plugin-api<|>PACKAGING=jar<|>POLL_VERSION_FROM=<|>POLL_VERSION_TO=<|>REPO_URL=https://repo1.maven.org/maven2
```
```json
[
  {
    "key": "GROUP_ID",
    "value": "cd.go.plugin"
  },
  {
    "key": "ARTIFACT_ID",
    "value": "go-plugin-api"
  },
  {
    "key": "PACKAGING",
    "value": "jar"
  },
  {
    "key": "POLL_VERSION_FROM",
    "value": ""
  },
  {
    "key": "POLL_VERSION_TO",
    "value": ""
  }
]
```
The string computed to generate fingerprint:
```
plugin-id=maven-repo<|>GROUP_ID=cd.go.plugin<|>ARTIFACT_ID=go-plugin-api<|>PACKAGING=jar<|>POLL_VERSION_FROM=null<|>POLL_VERSION_TO=null<|>REPO_URL=https://repo1.maven.org/maven2
```

The difference in the generated string is `POLL_VERSION_FROM=<|>POLL_VERSION_TO=<|>` and `POLL_VERSION_FROM=`**null**`<|>POLL_VERSION_TO=`**null**`<|>`

As you can see the difference in the fingerprint string, the ConfigurationProperty value is set to empty string when de-serialzing from XML but set to **null** for JSON.

I think either way we should ignore ConfigurationProperty with either null or empty string as value.

**Comment by maheshp** [comment]
Currently the following packages even though same are considered as unique because of the way fingerprint is generated.
```xml
<package id="6980a98d-bc50-4fac-b635-3c0890553e9f" name="pack1">
	  <configuration>
	    <property>
	      <key>GROUP_ID</key>
	      <value>cd.go.plugin</value>
	    </property>
	    <property>
	      <key>ARTIFACT_ID</key>
	      <value>go-plugin-api</value>
	    </property>
	    <property>
	      <key>PACKAGING</key>
	      <value>jar</value>
	    </property>
	  </configuration>
	</package>
	<package id="0febedc0-10f6-4658-a861-019a0d1bce10" name="pack2">
	  <configuration>
	    <property>
	      <key>GROUP_ID</key>
	      <value>cd.go.plugin</value>
	    </property>
	    <property>
	      <key>ARTIFACT_ID</key>
	      <value>go-plugin-api</value>
	    </property>
	    <property>
	      <key>PACKAGING</key>
	      <value>jar</value>
	    </property>
	    <property>
	      <key>POLL_VERSION_FROM</key>
	    </property>
	    <property>
	      <key>POLL_VERSION_TO</key>
	    </property>
	  </configuration>
  </package>
```

**Comment by GaneshSPatil** [comment]
Fixed as part of #7982

Verified on GoCD Version: `20.3.0 (11467-1fb7c3cc3a71abb71e827f5bbdc2cb139edcd660).`

The save fails with following error:
<img width="1087" alt="Screen Shot 2020-04-10 at 12 37 32 PM" src="https://user-images.githubusercontent.com/15275847/78970981-294c5c80-7b28-11ea-89b4-1aeb6b0210ac.png">
